### ISSUE TITLE
server product list mismatch in multi server setup

### ISSUE THREAD

**Issue by gyorb** [Ericsson/codechecker#1464]
Product list in the database is read up only once at server start. Managing this list it is done only if a product is added or removed on that specific server.
In a multi server setup this will not work properly because the other servers in memory product list will not be updated. Every server using the same config database needs to be restarted if a product is added or removed to update they product list.

Product configurations should be read from the database, if it causes to much extra queries we should introduce some product list invalidation time, so the product list will be updated on each server after some time automatically.

- [ ] #1464 partially solves the problem. If products are added on one server the other servers will query for the product if the product is not found in the cache.
- [ ] Updating the product list if a product is removed still needs to be solved.

**Comment by whisperity** [comment]
This has been definitely implemented as #1111. At the first moment someone wants to access a product that isn't seen by the server, either via `store` or by querying or opening the URL of the product endpoint directly, the server *should* (or at least, back in August it definitely did) try an extra round of finding the product in the shared config database.

**Comment by gyorb** [comment]
Yes but `get_product` checks only locally and if it is not found reports back to the user that there is no such product:
https://github.com/Ericsson/codechecker/blob/240bdc3716b2c32588d2017711a61fe0484ca37d/libcodechecker/server/server.py#L775-L779
`get_product` will return None and `__check_prod_db` will return a ValueError
https://github.com/Ericsson/codechecker/blob/240bdc3716b2c32588d2017711a61fe0484ca37d/libcodechecker/server/server.py#L257-L270

**Comment by gyorb** [comment]
There is still a problem when a product is removed the cache will not be cleaned up and it will be listed as an existing product on some servers even if the product database and the product config was removed.

**Comment by csordasmarton** [comment]
With the #3349 patch CodeChecker will use kind of "multi-server" setup, so we need to solve this problem after the patch is merged. For this we need to eliminate local caches for products and local sessions too.

**Comment by bruntib** [comment]
This issue works now.
